data/lua/core/unit_test.lua

```diff
@@ -0,0 +1,212 @@
+
+--[========[Unit Test Helpers]========]
+
+if rawget(_G, 'unit_test') ~= nil then
+	print("Loading unit_test module...")
+	
+	--! Causes the test to complete with a specific result.
+	--! This is mostly an internal detail but might very occasionally be useful to produce a result other than pass or fail
+	function unit_test.finish(result)
+		wesnoth.wml_actions.endlevel{
+			test_result = result,
+			linger_mode = true
+		}
+	end
+
+	--! End the test in unconditional failure
+	function unit_test.fail()
+		unit_test.finish('fail')
+	end
+
+	--! End the test in success if all assertions passed, otherwise in failure
+	function unit_test.pass()
+		unit_test.finish('pass')
+	end
+
+	--! Convert a value to a string for output
+	--! Strings are output quoted, all other types just use tostring
+	function unit_test.tostring(val)
+		-- This exists so custom behaviour can be added for specific types if required.
+		if type(val) == 'string' then
+			-- Strings are output quoted but without escaped_split
+			-- Pick a quote style that the string doesn't contain
+			if val:find("'") == nil then
+				return "'" .. val .. "'"
+			elseif val:find('"') == nil then
+				return '"' .. val .. '"'
+			else
+				local n = 0;
+				-- This loop is guaranteed to end eventually, although it could run a very large number of times on particularly pathological input.
+				while true do
+					local left = '[' .. string.rep('=', n) .. '['
+					local right = ']' .. string.rep('=', n) .. ']'
+					if val:find(left) == nil and val:find(right) == nil then
+						return left .. val .. right
+					end
+					n = n + 1
+				end
+			end
+		end
+		return tostring(val)
+	end
+
+	--! Output a log message to the test output
+	function unit_test.log(prefix, message)
+		std_print(prefix .. ': ' .. message)
+	end
+
+	--! End the test in failure if condition is true and success otherwise
+	--! Can be thought of as 'return not condition'
+	function unit_test.fail_if(condition, message)
+		if condition then
+			unit_test.log('Failed because fail_if condition was true', message)
+			unit_test.fail()
+		else
+			unit_test.pass()
+		end
+	end
+
+	--! End the test in success if condition is true and failure otherwise
+	--! Can be thought of as 'return condition'
+	function unit_test.pass_if(condition, message)
+		if condition then
+			unit_test.pass()
+		else
+			unit_test.log('Failed because pass_if condition was false', message)
+			unit_test.fail()
+		end
+	end
+
+	--! Fail the test with a message unless the condition is true
+	function unit_test.assert(condition, message)
+		if not condition then
+			unit_test.log('Assertion failed', message)
+			unit_test.fail()
+		end
+	end
+
+	--! Fail the test with a message unless the condition is false
```

Comment:
The reason to use this would be similar to the reason to use eg `assert_equal` - so that the negation can be included in the message if the assertion fails. Admittedly it's not _quite_ as useful for negation as it is for equality… do you think it's worth keeping?

Reply:
For new APIs, I'm strongly in the "less is better" camp.